python_modules/dagster/dagster/core/asset_defs/decorators.py

```diff
@@ -309,6 +313,154 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+    )
+    asset_keys_by_output_name = check.opt_dict_param(
+        asset_keys_by_output_name, "asset_keys_by_output_name", key_type=str, value_type=AssetKey
+    )
+    internal_asset_deps = check.opt_dict_param(
+        internal_asset_deps, "internal_asset_deps", key_type=str, value_type=set
+    )
+
+    if callable(name):
+        check.invariant(
+            isinstance(name, OpDefinition),
+            "assets_definition decorator can only be applied to an OpDefinition",
+        )
+        op_def = name
+        return AssetsDefinition(
+            input_names_by_asset_key=_infer_input_names_by_asset_key(
+                op_def,
+                asset_keys_by_input_name or {},
+            ),
+            output_names_by_asset_key=_infer_output_names_by_asset_key(
+                op_def, asset_keys_by_output_name or {}
+            ),
+            op=op_def,
+        )
+
+    def inner(op_def: OpDefinition) -> AssetsDefinition:
+        check.invariant(
+            isinstance(op_def, OpDefinition),
+            "assets_definition decorator can only be applied to an OpDefinition",
+        )
+        output_names_by_asset_key = _infer_output_names_by_asset_key(
+            op_def, asset_keys_by_output_name or {}
+        )
+        asset_key_by_output_name = {
+            output_name: asset_key for asset_key, output_name in output_names_by_asset_key.items()
+        }
+        transformed_internal_asset_deps = {}
+        if internal_asset_deps:
+            for output_name, asset_keys in internal_asset_deps.items():
+                check.invariant(
+                    output_name in asset_key_by_output_name,
+                    f"output_name {output_name} specified in internal_asset_deps does not exist in the decorated function",
+                )
+                transformed_internal_asset_deps[asset_key_by_output_name[output_name]] = asset_keys
+
+        return AssetsDefinition(
+            input_names_by_asset_key=_infer_input_names_by_asset_key(
+                op_def,
+                asset_keys_by_input_name or {},
+            ),
+            output_names_by_asset_key=output_names_by_asset_key,
+            op=op_def,
+            asset_deps=transformed_internal_asset_deps or None,
+        )
+
+    return inner
+
+
+def _get_input_param_names(fn_params: List[funcsigs.Parameter]) -> List[str]:
+    is_context_provided = len(fn_params) > 0 and fn_params[0].name in get_valid_name_permutations(
+        "context"
+    )
+    return [
+        input_param.name for input_param in (fn_params[1:] if is_context_provided else fn_params)
+    ]
+
+
+def _infer_input_names_by_asset_key(
+    op_def: OpDefinition, asset_keys_by_input_name: Mapping[str, AssetKey]
+) -> Mapping[AssetKey, str]:
+    # Infer non-argument deps for inputs with type In(nothing) with AssetKey(input_name)
+
+    input_param_names = []
+    if isinstance(op_def.compute_fn, DecoratedSolidFunction):
```

Comment:
in what scenarios would this condition be false?

Reply:
Invoking the decorated op when the op is not constructed using the @op decorator. I can add raise an error when this occurs